### ISSUE TITLE
funannotate util gff2tbl - Error, can't find ID or Parent. Malformed GFF file.

### ISSUE THREAD

**Issue by nextjen92** [nextgenusfs/funannotate#581]
I am trying to convert a GFF3 file to a GenBank feature table so that I can correct some partial gene predictions in the table then convert to sequin for submission, rather than convert GFF to sequin and then try to fix. 

However, when I run "funannotate util gff2tbl" I get the following error:
Error, can't find ID or Parent. Malformed GFF file.
scaffold10005   maker   mRNA    781     3662    .       -       .       ID=Hamer_029131_mRNA;  Parent=Hamer_G029131; Name=Cdcp-L3

As you can see, there is "ID" and "Parent" in the GFF file. The GFF has been validated withhttp://genometools.org/cgi-bin/gff3validator.cgi (recommended by NCBI for GFF annotations). 

Why is it not recognizing the ID and Parent in my GFF and how can I make it?

**Comment by nextgenusfs** [comment]
Perhaps just a `grep "Hamer_G029131" yourfile.gff` will be sufficient to see the problem with this specific model.  Its probably the order in which the lines are written, ie the gene feature should be the first line of any gene model (gene --> mRNA --> exon/CDS). I think I have code in here that can sort GFF3 in the correct way, so maybe that needs to be done by default in this script.

**Comment by nextgenusfs** [comment]
Hmm, that all looks fine to me on the surface.  What version of funannotate are you using?

Would you be able to share the genome and GFF so we can test locally?

**Comment by nextjen92** [comment]
I've attached a subset of gene features for 50 scaffolds and the fasta with those scaffolds' sequences. 

[jmp_gff2tbl-test.gff.txt](https://github.com/nextgenusfs/funannotate/files/6238947/jmp_gff2tbl-test.gff.txt)
[jmp_gff2tbl-test.fa.txt](https://github.com/nextgenusfs/funannotate/files/6238948/jmp_gff2tbl-test.fa.txt)

**Comment by nextgenusfs** [comment]
Hi @nextjen92 I'm not getting any errors with these files, so I'm thinking has something to do with either the version of fun annotate you are running or your install.
```
(py3-funannotate) $ funannotate util gff2tbl -g jmp_gff2tbl-test.gff -f jmp_gff2tbl-test.fa > jmp_gff2tbl-test.tbl
(py3-funannotate) $ funannotate --version
funannotate v1.8.5
```

[jmp_gff2tbl-test.tbl.txt](https://github.com/nextgenusfs/funannotate/files/6244917/jmp_gff2tbl-test.tbl.txt)

**Comment by nextjen92** [comment]
I tested the files I sent yesterday just now, and they appear to work with the exception that all of my product information shows up as "hypothetical protein" rather than the annotations in the GFF. 

However, now when I try to run the full fasta and GFF, I get a new error:
`Traceback (most recent call last):
  File "/data/app/conda/envs/funannotate/bin/funannotate", line 10, in <module>
    sys.exit(main())
  File "/data/app/conda/envs/funannotate/lib/python3.6/site-packages/funannotate/funannotate.py", line 705, in main
    mod.main(arguments)
  File "/data/app/conda/envs/funannotate/lib/python3.6/site-packages/funannotate/utilities/gff2tbl.py", line 335, in main
    Genes = lib.gff2dict(args.gff3, args.fasta, Genes)
  File "/data/app/conda/envs/funannotate/lib/python3.6/site-packages/funannotate/library.py", line 4235, in gff2dict
    contig, source, feature, start, end, score, strand, phase, attributes = line.split('\t')
ValueError: not enough values to unpack (expected 9, got 1)`

I am using v1.8.5.

**Comment by nextgenusfs** [comment]
GFF3 is problematic because all the information is in column 9, its impossible to build a parser for all the different flavors out there.  The product deflines aren't coming through because funannotate is expecting them to be located on the `mRNA` feature per GFF3 spec, for example:
```
scaffold1024	maker	gene	119140	154462	.	+	.	ID=Hamer_G009739;Name=GET4-L;Product=Golgi to ER traffic protein 4-like;locus_tag=Hamer_G009739
scaffold1024	maker	mRNA	119140	154462	.	+	.	ID=Hamer_009739_mRNA;Parent=Hamer_G009739;Name=GET4-L
scaffold1024	maker	exon	119140	119217	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	122471	122565	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	122811	122947	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	135900	136038	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	137642	137788	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	147523	147674	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	exon	151919	154462	.	+	.	ID=Hamer_009739_exon;Parent=Hamer_009739_mRNA
scaffold1024	maker	CDS	119140	119217	.	+	0	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	122471	122565	.	+	0	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	122811	122947	.	+	1	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	135900	136038	.	+	2	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	137642	137788	.	+	1	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	147523	147674	.	+	1	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
scaffold1024	maker	CDS	151919	152034	.	+	2	ID=Hamer_009739;Parent=Hamer_009739_mRNA;Name=GET4-L;Product=Golgi to ER traffic protein 4-like
```
The `Product=` is on the `gene` and `CDS` features but the parser in funannotate here is looking for that information on the `mRNA` feature. This is because genes can have more than 1 transcript, so product deflines should be on the mRNA feature.

The error you are getting is because the script died on parsing a line that it doesn't expect to see -- so there is some additional lines that don't have the expected 8 tabs in them. You could pull out all lines with less than 8 tabs in your file with this awk script and then clean it up:
```
awk -F'\t' 'NF<=8 {print}' your_file.gff3
```

**Comment by nextgenusfs** [comment]
And if you let me know which lines were in the GFF file that were causing it to fail I can add to the code to ignore those.

**Comment by nextjen92** [comment]
After a bit of digging, I discovered a few lines in the GFF that had spaces instead of tabs delimiting the 9 columns - they were very far down in the GFF so not included in the 50 scaffold subset that I sent for testing. I'm not sure how that happened, but after manually replacing the spaces with tabs, it worked great!

I don't understand what caused the original "can't find ID or Parent" error - it seemed to resolve itself overnight, only to give a new error that was in fact a format problem in my GFF. The second error was due to the spaces and all it running well now.
